torch/overrides.py

```diff
@@ -1852,30 +1842,55 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
 
     def __enter__(self):
         old = _get_torch_function_mode()
-        if hasattr(self, "inner"):
-            raise RuntimeError(f"{self} has already been used as a mode. Please use a fresh version or use restore")
-        else:
-            self.inner = old
-            if old is None:
-                self.ancestors = set()
-            else:
-                self.ancestors = self.inner.ancestors.union({self.inner})
-        _set_torch_function_mode(self)
+        if self in _cur_torch_function_mode or self is old:
+            raise RuntimeError(f"{self} is already active in the mode stack")
+        _push_mode(self)
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        _set_torch_function_mode(self.inner)
-
-    @contextlib.contextmanager
-    def restore(self):
-        return _restore_mode(self, mode_info=_TorchFunctionModeInfo())
+        _pop_mode()
 
     @classmethod
     def push(cls, *args, **kwargs):
         warnings.warn("`Mode.push()` is no longer necessary and can be replaced with just `with Mode()`")
         instance = cls(*args, **kwargs)
         return instance
 
+
+def _get_current_function_mode():
+    return _cur_torch_function_mode[-1] if len(_cur_torch_function_mode) > 0 else None
+
+
+def _push_mode(mode):
+    _set_torch_function_mode(_TorchFunctionStackMode())
```

Comment:
We should only have to do this on the 0 to 1 transition?

Reply:
Done